lms/djangoapps/commerce/utils.py

```diff
@@ -109,3 +117,247 @@ def upgrade_url(self, user, course_key):
+        enrollee.username,
+        entitlement_uuid
+    )
+
+    try:
+        refund_ids = api_client.refunds.post(
+            {
+                'order_number': course_entitlement.order_number,
+                'username': enrollee.username,
+                'entitlement_uuid': entitlement_uuid,
+            }
+        )
+    except Exception as exc:  # pylint: disable=broad-except
+        # Catch any possible exceptions from the Ecommerce service to ensure we fail gracefully
+        log.exception(
+            "Unexpected exception while attempting to initiate refund for user [%s], "
+            "course entitlement [%s] message: [%s]",
+            course_entitlement.user.id,
+            course_entitlement.uuid,
+            str(exc)
+        )
+        return False
+
+    if refund_ids:
+        log.info(
+            'Refund successfully opened for user [%s], course entitlement [%s]: %r',
+            enrollee.username,
+            entitlement_uuid,
+            refund_ids,
+        )
+
+        return _process_refund(
+            refund_ids=refund_ids,
+            api_client=api_client,
+            course_product=course_entitlement
+        )
+    else:
+        log.info('No refund opened for user [%s], course entitlement [%s]', enrollee.id, entitlement_uuid)
+        return True
+
+
+def refund_seat(course_enrollment):
+    """
+    Attempt to initiate a refund for any orders associated with the seat being unenrolled,
+    using the commerce service.
+
+    Arguments:
+        course_enrollment (CourseEnrollment): a student enrollment
+
+    Returns:
+        A list of the external service's IDs for any refunds that were initiated
+            (may be empty).
+
+    Raises:
+        exceptions.SlumberBaseException: for any unhandled HTTP error during communication with the E-Commerce Service.
+        exceptions.Timeout: if the attempt to reach the commerce service timed out.
+    """
+    User = get_user_model()  # pylint:disable=invalid-name
+    course_key_str = unicode(course_enrollment.course_id)
+    enrollee = course_enrollment.user
+
+    service_user = User.objects.get(username=settings.ECOMMERCE_SERVICE_WORKER_USERNAME)
+    api_client = ecommerce_api_client(service_user)
+
+    log.info('Attempting to create a refund for user [%s], course [%s]...', enrollee.id, course_key_str)
+
+    refund_ids = api_client.refunds.post({'course_id': course_key_str, 'username': enrollee.username})
+
+    if refund_ids:
+        log.info('Refund successfully opened for user [%s], course [%s]: %r', enrollee.id, course_key_str, refund_ids)
+
+        _process_refund(
+            refund_ids=refund_ids,
+            api_client=api_client,
+            course_product=course_enrollment,
+        )
+    else:
+        log.info('No refund opened for user [%s], course [%s]', enrollee.id, course_key_str)
+
+    return refund_ids
+
+
+def _process_refund(refund_ids, api_client, course_product):
```

Comment:
Note: Method used by both refund entitlement and refund seat. Copied over from Signals.

Reply:
I'm not a huge fan of the duck typing being applied to `course_product`. It seems like it can be either a `CourseEnrollment` or a `CourseEntitlement`, which are very different types of objects. There are some places in here where the codepath diverges depending on which type of `course_product` was passed, and the method we use to determine which path to take is brittle (presence of `uuid` property).

I would prefer to be explicit about the values passed to this method, either via keyword args for `course_entitlement`/`course_enrollment` or via explicit arguments for the values the method actually needs (`user`, `mode`, `course_id`, `is_entitlement`).